pkg/server/plugin/nodeattestor/azure/msi.go

```diff
@@ -0,0 +1,193 @@
+package azure
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/hcl"
+	"github.com/spiffe/spire/pkg/common/jwtutil"
+	"github.com/spiffe/spire/pkg/common/plugin/azure"
+	spi "github.com/spiffe/spire/proto/common/plugin"
+	"github.com/spiffe/spire/proto/server/nodeattestor"
+	"github.com/zeebo/errs"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+const (
+	pluginName = "azure_msi"
+
+	tokenLeeway           = time.Minute * 5
+	keySetRefreshInterval = time.Hour
+	azureConfigURL        = "https://login.microsoftonline.com/common/.well-known/openid-configuration"
+)
+
+var (
+	msiError = errs.Class("azure-msi")
+)
+
+type TenantConfig struct {
+	ResourceID string `hcl:"resource_id"`
+}
+
+type MSIAttestorConfig struct {
+	TrustDomain string                   `hcl:"trust_domain"`
+	Tenants     map[string]*TenantConfig `hcl:"tenants"`
+}
+
+type MSIAttestorPlugin struct {
+	mu     sync.RWMutex
+	config *MSIAttestorConfig
+
+	hooks struct {
+		now            func() time.Time
+		keySetProvider jwtutil.KeySetProvider
+	}
+}
+
+var _ nodeattestor.Plugin = (*MSIAttestorPlugin)(nil)
+
+func NewMSIAttestorPlugin() *MSIAttestorPlugin {
+	p := &MSIAttestorPlugin{}
+	p.hooks.now = time.Now
+	p.hooks.keySetProvider = jwtutil.NewCachingKeySetProvider(jwtutil.KeySetConfigURL(azureConfigURL), keySetRefreshInterval)
+	return p
+}
+
+func (p *MSIAttestorPlugin) Attest(stream nodeattestor.Attest_PluginStream) error {
+	req, err := stream.Recv()
+	if err != nil {
+		return msiError.Wrap(err)
+	}
+
+	config, err := p.getConfig()
+	if err != nil {
+		return err
+	}
+
+	if req.AttestedBefore {
+		return msiError.New("node has already attested")
+	}
+
+	if req.AttestationData == nil {
+		return msiError.New("missing attestation data")
+	}
+
+	if dataType := req.AttestationData.Type; dataType != pluginName {
+		return msiError.New("unexpected attestation data type %q", dataType)
+	}
+
+	if req.AttestationData.Data == nil {
+		return msiError.New("missing attestation data payload")
+	}
+
+	attestationData := new(azure.MSIAttestationData)
+	if err := json.Unmarshal(req.AttestationData.Data, attestationData); err != nil {
+		return msiError.New("failed to unmarshal data payload: %v", err)
+	}
+
+	if attestationData.Token == "" {
+		return msiError.New("missing token from attestation data")
+	}
+
+	keySet, err := p.hooks.keySetProvider.GetKeySet(stream.Context())
+	if err != nil {
+		return msiError.New("unable to obtain JWKS: %v", err)
+	}
+
+	token, err := jwt.ParseSigned(attestationData.Token)
+	if err != nil {
+		return msiError.New("unable to parse token: %v", err)
+	}
+
+	keyID, ok := getTokenKeyID(token)
+	if !ok {
+		return msiError.New("token missing key id")
+	}
+
+	keys := keySet.Key(keyID)
+	if len(keys) == 0 {
+		return msiError.New("key id %q not found", keyID)
+	}
+
+	claims := new(azure.MSITokenClaims)
+	if err := token.Claims(&keys[0], claims); err != nil {
```

Comment:
Hmmm I think we might be opening ourself up to a vulnerability here in which the alg is swapped for HS256? The discovery document specifies `"id_token_signing_alg_values_supported":["RS256"]`. We should assert that the alg value is RS256 as part of validation - either hardcoded or parsed out of the discovery doc (the former doesn't seem _too_ unreasonable).... not sure if the jwt lib we're using does any magic to try to prevent this

A test explicitly exercising this would be great

Reply:
Yup, good call. The library will fail signature verification in this case (because the key type wouldn't match), but I agree it's better to be cautious. I'll either hardcode RS256 or pull it from the doc.